File: go/lesson01/solution/hello.go
package main

import (
	"context"
	"fmt"
	"log"
	"os"
	"time"

	"github.com/legosandorigami/opentelemetry-tutorial/lib/tracing"
	"go.opentelemetry.io/otel/attribute"
	"go.opentelemetry.io/otel/trace"
)

func main() {
	if len(os.Args) != 2 {
		panic("ERROR: Expecting one argument")
	}

	// initializing the OpenTelemetry TracerProvider with the service name "hello-world"
	tracerProvider, err := tracing.InitTracerProvider("hello-world")
	if err != nil {
		log.Fatalf("failed to create otel exporter: %v", err)
	}

	// creating a context and defering the shutdown of the TracerProvider to ensure proper cleanup
	ctx := context.Background()
	defer func() {
		if err := tracerProvider.Shutdown(ctx); err != nil {
			log.Fatalf("failed to shutdown TracerProvider: %v", err)
		}
	}()

	// getting a tracer from the tracer provider
	tracer := tracerProvider.Tracer("say-hello-tracer")

	helloTo := os.Args[1]

	// starting a new span named "say-hello"
	ctx, span := tracer.Start(ctx, "say-hello")

	// adding an attribute to the span
	span.SetAttributes(attribute.String("hello-to", helloTo))

	helloStr := fmt.Sprintf("Hello, %s!", helloTo)

	// adding logs to the span
	span.AddEvent("event", trace.WithAttributes(attribute.String("println", fmt.Sprintf("string-format: %s", helloStr))),
		trace.WithTimestamp(time.Now()),
	)

	println(helloStr)

	tracing.PrintSpanContents(span)

	// ending the span
	span.End()
}

File: python/lesson01/solution/hello.py
import sys
import time
from opentelemetry import trace
from opentelemetry.sdk.trace import TracerProvider, Tracer
from lib.tracing import init_tracer_provider

def say_hello(hello_to):
    # retrieve the global tracer provider
    tracer_provider: TracerProvider = trace.get_tracer_provider()

    # obtain a tracer instance from the tracer provider
    tracer: Tracer = tracer_provider.get_tracer("say_hello_tracer")
    
    # starting a new span for the 'say-hello' operation
    with tracer.start_as_current_span('say-hello') as span:
        # seting an attribute on the span
        span.set_attribute("hello-to", hello_to)

        hello_str = f'Hello, {hello_to}!'
        
        # adding an event to the span indicating that the string has been formatted
        span.add_event('string-format-event', {'string-formatted': hello_str})

        # printing the greeting
        print(hello_str)
        
        # adding an event to the span indicating that the greeting has been printed
        span.add_event('print-event', {'println': True})

        print(span.get_span_context())

def main():
    # ensuring the program is called with one argument
    assert len(sys.argv) == 2

    # setting and retrieving the global tracer provider
    tracer_provider: TracerProvider = init_tracer_provider('hello-world')

    hello_to = sys.argv[1]
    
    say_hello(hello_to)
    
    # Optional: yield to IOLoop to flush the spans
    # time.sleep(2)  # Can be removed if shutdown() is sufficient
    
    # shutting down the tracer provider to ensure all the spans are exported
    tracer_provider.shutdown()

if __name__ == '__main__':
    main()
